File: frontend/src/api/codegen.ts
import { api } from "@/api/client";
import type { FlowEdge, FlowNode } from "@/types";

export interface CodegenResponse {
  code: string;
  language: string;
  file_extension: string;
  warnings: string[];
}

export async function fetchSupportedLanguages(): Promise<string[]> {
  const { data } = await api.get<{ languages: string[] }>("/codegen/languages");
  return data.languages;
}

export async function generateCodeFromFlowchart(
  nodes: FlowNode[],
  edges: FlowEdge[],
  language: string
): Promise<CodegenResponse> {
  const { data } = await api.post<CodegenResponse>("/codegen/generate", {
    nodes: nodes.map((n) => ({ id: n.id, type: n.type, text: n.text })),
    edges: edges.map((e) => ({ id: e.id, fromNodeId: e.fromNodeId, toNodeId: e.toNodeId, label: e.label ?? null })),
    language,
  });
  return data;
}

File: api/app/api/routes/codegen.py
from fastapi import APIRouter, Depends, HTTPException

from app.api.deps import get_current_user
from app.codegen.emitters.registry import EMITTER_REGISTRY
from app.codegen.pipeline import generate_code
from app.codegen.schemas import CodegenRequest, CodegenResponse, SupportedLanguagesResponse
from app.models.user import User

router = APIRouter(prefix="/codegen", tags=["codegen"])


@router.get("/languages", response_model=SupportedLanguagesResponse)
def list_languages():
    return SupportedLanguagesResponse(languages=sorted(EMITTER_REGISTRY.keys()))


@router.post("/generate", response_model=CodegenResponse)
def generate(payload: CodegenRequest, _: User = Depends(get_current_user)):
    try:
        result = generate_code(
            [n.model_dump() for n in payload.nodes],
            [e.model_dump() for e in payload.edges],
            payload.language,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return CodegenResponse(
        code=result.code,
        language=result.language,
        file_extension=result.file_extension,
        warnings=result.warnings,
    )
